### PR TITLE
feat: use epoch schedule for uptime calculation

### PR DESCRIPTION
**Pull request by sunnygleason**
PROBLEM:

* we were using hard-coded slots_per_epoch

SOLUTION:

* use the new getEpochSchedule API

### CODE REVIEW

File: api/uptime-crawler.js [hunk 1]
```diff
@@ -20,12 +20,51 @@ function getClient() {
...
+  if (!epochSchedule.warmup || epoch >= epochSchedule.first_normal_epoch) {
+    return epochSchedule.slots_per_epoch;
+  }
+
+  return null;
+}
+
+function getUptime(epochSchedule, voteState, lat, ts) {
+  const uptime = _.reduce(
+    voteState.epochCredits,
```
Comment (CriesofCarrots) [inline api/uptime-crawler.js]: What if a validator's voteState doesn't start at the first_normal_epoch? Won't that make their uptime much higher for the stage than it should be?
I think what you have is correct for mainnet, but for TdS we need to start at first_normal_epoch regardless of when the validator actually started voting.

Comment (sunnygleason) [reply api/uptime-crawler.js]: @CriesofCarrots that's a great point, I'll look at this right now to try to sort all the edge cases out (or at least as many as possible)!

**Comment by sunnygleason** [comment]
@CriesofCarrots ready for review, thank you!

File: api/uptime-crawler.js [hunk 2]
```diff
@@ -20,12 +20,52 @@ function getClient() {
...
+
+      if (!slotsInEpoch) {
+        return a;
+      }
+
+      const credits = v.credits - v.prevCredits;
+
+      a.unshift({
+        epoch: v.epoch,
+        credits_earned: credits,
```
Comment (CriesofCarrots) [inline api/uptime-crawler.js]: nit: Any reason these fields need to be snake case?

**Comment by sunnygleason** [comment]
> Looking forward to seeing this in action!
> I don't see any handling to start TdS uptime from the "magic" start epoch, though (epoch 10). Am I missing it?

good question, will follow up in slack...

**Pull request merged**